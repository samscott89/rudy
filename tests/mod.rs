use rust_debuginfo::*;

use rstest::rstest;
use rstest_reuse::{self, *};

#[macro_export]
macro_rules! function_name {
    () => {{
        // Okay, this is ugly, I get it. However, this is the best we can get on a stable rust.
        fn f() {}
        fn type_name_of<T>(_: T) -> &'static str {
            std::any::type_name::<T>()
        }
        let name = type_name_of(f);
        // `3` is the length of the `::f`.
        &name[..name.len() - 3]
    }};
}

#[macro_export]
macro_rules! setup {
    ($target:ident) => {
        let _ = tracing_subscriber::fmt::try_init();
        let mut settings = insta::Settings::clone_current();

        // get current OS as a prefix
        settings.set_snapshot_suffix($target);
        settings.set_prepend_module_to_snapshot(false);

        let _guard = settings.bind_to_scope();
        let test_name = crate::function_name!();
        let test_name = test_name
            .strip_prefix("rust_debuginfo::")
            .unwrap_or(test_name);
        let test_name = test_name.strip_prefix("tests::").unwrap_or(test_name);
        let _span = tracing::info_span!("test", test_name, $target).entered();
    };
}

#[template]
#[rstest]
#[case("aarch64-unknown-linux-gnu")]
#[case("x86_64-unknown-linux-gnu")]
#[case("aarch64-apple-darwin")]
#[case("x86_64-apple-darwin")]
pub fn binary_target(#[case] target: &str) {}

pub fn binary_path(target: &str) -> String {
    let workspace_dir = std::env::var("CARGO_WORKSPACE_DIR").unwrap();
    format!("{workspace_dir}target/{target}/debug-test/simple-test")
}

#[apply(binary_target)]
fn test_resolve_function(#[case] target: &str) {
    setup!(target);
    let path = binary_path(target);
    let db = crate::DebugDb::new().unwrap();
    let resolver = DebugInfo::new(&db, &path).unwrap();

    insta::assert_debug_snapshot!(resolver.resolve_function("main").unwrap());
    insta::assert_debug_snapshot!(resolver.resolve_function("function_call").unwrap());
}

#[apply(binary_target)]
fn test_resolve_position(#[case] target: &str) {
    setup!(target);
    let path = binary_path(target);
    let db = crate::DebugDb::new().unwrap();
    let resolver = DebugInfo::new(&db, &path).unwrap();

    // should be the position of the `let y = x + 1;` line
    let addrs = resolver
        .resolve_position("examples/simple-test/src/main.rs", 4, None)
        .unwrap()
        .unwrap();
    insta::assert_debug_snapshot!(addrs);
    assert_eq!(
        resolver.address_to_line(addrs.address).unwrap(),
        ResolvedLocation {
            function: "function_call".to_string(),
            file: "examples/simple-test/src/main.rs".to_string(),
            line: 4,
        }
    );

    // should be the position of the `const Z: u64 = 0xdeadbeef;` line
    // however, this gets inlined and so we end up at the next line
    //
    // same behaviour as LLDB:
    //   (lldb) b simple-test.rs:21
    //   Breakpoint 4: where = simple-test`simple_test::main::h8787f5d764ea460c + 20 at simple-test.rs:22:5, address = 0x00000001000041a0``
    let addrs = resolver
        .resolve_position("examples/simple-test/src/main.rs", 16, None)
        .unwrap()
        .unwrap();
    // address of line 17
    insta::assert_debug_snapshot!(addrs);

    assert_eq!(
        resolver.address_to_line(addrs.address).unwrap(),
        ResolvedLocation {
            function: "main".to_string(),
            file: "examples/simple-test/src/main.rs".to_string(),
            // This is right: it resolves to the next line cause
            // that's closest line that has an instruction
            line: 17,
        }
    );

    // TODO: come up with a way to test this (stack frames)
    // assert_eq!(
    //     resolver.address_to_line(0x100003948).unwrap(),
    //     ResolvedLocation {
    //         function: "main".to_string(),
    //         file: "examples/simple-test/src/main.rs".to_string(),
    //         // TODO(Sam): this is wrong. This should be line 18
    //         // based on the output of LLDB.
    //         // this is currently pointing to the end of the function rather
    //         // than the function call.
    //         // suspect we need to track "end of statement" smarter or something
    //         line: 19,
    //     }
    // );
}

#[apply(binary_target)]
fn test_load_file(#[case] target: &str) {
    setup!(target);

    let path = binary_path(target);
    let db = crate::DebugDb::new().unwrap();
    let parsed = DebugInfo::new(&db, &path).unwrap();

    insta::assert_debug_snapshot!(parsed);
}
