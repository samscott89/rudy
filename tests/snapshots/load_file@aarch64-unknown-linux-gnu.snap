---
source: tests/mod.rs
expression: parsed
snapshot_kind: text
---
DebugInfo {
    debug_files: [
        DebugFile {
            [salsa id]: Id(800),
            file: File {
                [salsa id]: Id(0),
                path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                member_file: None,
                mtime: SystemTime {
                    tv_sec: 1748701990,
                    tv_nsec: 341906278,
                },
                size: 3829072,
            },
            relocatable: false,
        },
    ],
    index: Index {
        [salsa id]: Id(2400),
        data: IndexData {
            name_to_file: {
                NameId {
                    path: [
                        "simple_test",
                    ],
                    name: "function_call",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                    ],
                    name: "main",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                    ],
                    name: "report",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                    ],
                    name: "precondition_check",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                        "std",
                        "rt",
                        "lang_start",
                    ],
                    name: "{closure#0}<()>",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                        "std",
                        "rt",
                        "lang_start",
                        "core",
                        "fmt",
                        "rt",
                        "core",
                        "fmt",
                        "rt",
                        "alloc",
                        "collections",
                        "btree",
                        "navigate",
                        "LazyLeafRange",
                        "panic_abort",
                    ],
                    name: "__rust_panic_cleanup",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                        "std",
                        "rt",
                        "lang_start",
                        "core",
                        "fmt",
                        "rt",
                        "core",
                        "fmt",
                        "rt",
                        "alloc",
                        "collections",
                        "btree",
                        "navigate",
                        "LazyLeafRange",
                        "panic_abort",
                    ],
                    name: "__rust_start_panic",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                        "std",
                        "rt",
                        "lang_start",
                        "core",
                        "fmt",
                        "rt",
                        "core",
                        "fmt",
                        "rt",
                        "alloc",
                        "collections",
                        "btree",
                        "navigate",
                        "LazyLeafRange",
                        "panic_abort",
                        "__rust_start_panic",
                    ],
                    name: "abort",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                        "std",
                        "rt",
                        "lang_start",
                        "core",
                        "fmt",
                        "rt",
                        "core",
                        "fmt",
                        "rt",
                        "alloc",
                        "collections",
                        "btree",
                        "navigate",
                        "LazyLeafRange",
                        "panic_abort",
                        "__rust_start_panic",
                        "core",
                        "fmt",
                        "Formatter",
                        "core",
                        "num",
                        "dec2flt",
                        "decimal_seq",
                        "DecimalSeq",
                    ],
                    name: "left_shift",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                        "std",
                        "rt",
                        "lang_start",
                        "core",
                        "fmt",
                        "rt",
                        "core",
                        "fmt",
                        "rt",
                        "alloc",
                        "collections",
                        "btree",
                        "navigate",
                        "LazyLeafRange",
                        "panic_abort",
                        "__rust_start_panic",
                        "core",
                        "fmt",
                        "Formatter",
                        "core",
                        "num",
                        "dec2flt",
                        "decimal_seq",
                        "DecimalSeq",
                    ],
                    name: "right_shift",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
                NameId {
                    path: [
                        "simple_test",
                        "core",
                        "ptr",
                        "alignment",
                        "std",
                        "process",
                        "{impl#59}",
                        "core",
                        "hint",
                        "core",
                        "alloc",
                        "layout",
                        "{impl#0}",
                        "from_size_align_unchecked",
                        "std",
                        "rt",
                        "lang_start",
                        "core",
                        "fmt",
                        "rt",
                        "core",
                        "fmt",
                        "rt",
                        "alloc",
                        "collections",
                        "btree",
                        "navigate",
                        "LazyLeafRange",
                        "panic_abort",
                        "__rust_start_panic",
                        "core",
                        "fmt",
                        "Formatter",
                        "core",
                        "num",
                        "dec2flt",
                        "decimal_seq",
                        "DecimalSeq",
                        "compiler_builtins",
                        "int",
                        "udiv",
                        "__udivti3",
                    ],
                    name: "__udivti3",
                }: DebugFile {
                    [salsa id]: Id(800),
                    file: File {
                        [salsa id]: Id(0),
                        path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                        member_file: None,
                        mtime: SystemTime {
                            tv_sec: 1748701990,
                            tv_nsec: 341906278,
                        },
                        size: 3829072,
                    },
                    relocatable: false,
                },
            },
            source_to_file: {},
            base_address: {
                NameId {
                    path: [
                        "simple_test",
                    ],
                    name: "function_call",
                }: 25812,
                NameId {
                    path: [
                        "simple_test",
                    ],
                    name: "main",
                }: 26124,
            },
            address_info: AddressTree {
                root: Some(
                    Node {
                        info: FunctionAddressInfo {
                            start: 26124,
                            end: 26188,
                            file: DebugFile {
                                [salsa id]: Id(800),
                                file: File {
                                    [salsa id]: Id(0),
                                    path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                                    member_file: None,
                                    mtime: SystemTime {
                                        tv_sec: 1748701990,
                                        tv_nsec: 341906278,
                                    },
                                    size: 3829072,
                                },
                                relocatable: false,
                            },
                            name: NameId {
                                path: [
                                    "simple_test",
                                ],
                                name: "main",
                            },
                        },
                        max_end_in_subtree: 26188,
                        left: Some(
                            Node {
                                info: FunctionAddressInfo {
                                    start: 25812,
                                    end: 26124,
                                    file: DebugFile {
                                        [salsa id]: Id(800),
                                        file: File {
                                            [salsa id]: Id(0),
                                            path: "/Users/sam/work/tardis/target/aarch64-unknown-linux-gnu/debug-test/simple-test",
                                            member_file: None,
                                            mtime: SystemTime {
                                                tv_sec: 1748701990,
                                                tv_nsec: 341906278,
                                            },
                                            size: 3829072,
                                        },
                                        relocatable: false,
                                    },
                                    name: NameId {
                                        path: [
                                            "simple_test",
                                        ],
                                        name: "function_call",
                                    },
                                },
                                max_end_in_subtree: 26124,
                                left: None,
                                right: None,
                            },
                        ),
                        right: None,
                    },
                ),
            },
        },
    },
}
